✨ feat : 메시지 내역 조회 API 기능 구현 및 테스트 커버리지 향상

BE12-OZ opened the pull request:
## ✅ PR 요약
- 관련 이슈 번호: #125 
- 작업 요약: 메시지 내역 조회 API 구현 + is_read 처리 로직 추가 / mypy channels 예외 처리 설정 / 테스트 커버리지 향상
- 최신 작업 내역 추가사항에 작성
- 기존 feature/chat-list-new 브랜치 데이터 + 코치님 피드백 반영
- feature/chat-list-new 브랜치의 경우 많은 수정이 필요하고 오류가 발생하기 때문에 새 브랜치로 작업
- 오래된 베이스 커밋이력을 깔끔하게 정리하기 위해 새 브랜치에 체리픽(시간순)으로 깔끔하게 정리 

## 📄 상세 내용
- [x] 메시지 내역 조회 API :

•	기능 :
•	특정 스터디 그룹의 채팅 메시지 내역을 조회
•	페이지네이션 정책
•	1페이지(초기 로드): 기본 반환 개수 300개
•	이후(무한 스크롤): page_size 쿼리로 페이지당 100개 단위 추가 조회 가능(클라이언트가 page_size로 제어)
•	응답 항목: 메시지 id, content, created_at 등 기본 필드와 sender_id, sender_nickname, 그리고 현재 로그인한 사용자 기준의 is_read(읽음 여부)
•	보안: 인증된 사용자만 접근 가능, 해당 스터디 그룹의 멤버가 아니면 403 Forbidden
•	테스트: 인증/권한 검증, 페이지네이션(초기 300개, 이후 페이지), is_read 계산, sender_nickname 반환 등을 포함한 포괄적 테스트 추가 및 통과

- [x] 변경 파일 및 핵심 변경사항:

•	apps/chat/view.py :
•	ChatMessageListView (ListAPIView) 구현
•	extend_schema로 Summary/Description/Parameters/Responses 상세화 (page_size 예시 포함 — 300(초기), 100(무한스크롤)
•	권한(인증 + 스터디 그룹 멤버십) 검증 로직 포함

•	apps/chat/pagination.py :
•	ChatMessagePagination (DRF PageNumberPagination 상속)
•	기본 page_size = 300
•	page_size_query_param = "page_size"
•	DRF 표준 방식으로 paginate_queryset 오버라이드하여 mypy 타입 오류 방지

•	apps/chat/serializers.py :
•	ChatMessageSerializer
•	sender_nickname 및 is_read를 SerializerMethodField로 구현
•	get_is_read는 LastReadMessage 모델을 조회해 로그인 사용자의 읽음 상태를 정확히 반환
	
•	apps/chat/tests/test_chat_api.py :
•	ChatMessageListAPIViewTest 추가
•	범위: 비인증/비멤버 접근 차단, 페이지네이션(초기 300개, 다음 페이지 사이즈), is_read 정확성, sender_nickname 반환 등
	
•	코드 품질
•	mypy, black, isort 통과 보고.
•	테스트 스위트 전체 통과.

- [x] mypy 관련 문제 진단 및 해결 (정확한 원인·해결 기록)

문제 원인
•	channels 패키지에 공식 타입 스텁(stubs)이 포함되어 있지 않음 → mypy가 import-untyped 오류를 발생시킴
•	임시로 # type: ignore[import-untyped]를 import 라인에 추가했으나, black/isort가 임포트 포맷을 변경하면서 주석 위치가 바뀌어 mypy가 unused-ignore 오류를 추가로 보고함
•	포맷터(black/isort)와 정적분석기(mypy)의 주석 해석 방식 충돌이 발생

선택한 해결책(권장·적용된 방식)
•	코드 주석으로 개별적으로 무시하는 방식 대신, pyproject.toml에서 mypy 설정으로 예외 처리(전역 설정) 적용
•	이 설정을 [tool.mypy] 블록 아래에 추가하여 channels 관련 모든 import-untyped 오류를 무시
•	구체 설정(파일에 추가된 내용):

[[tool.mypy.overrides]]
module = "channels.*"
ignore_missing_imports = true

•	결과:
•	코드에 흩어진 # type: ignore[import-untyped] 주석을 제거 (정리함)
•	black/isort의 포맷 변경과 무관하게 mypy가 안정적으로 동작

추가 정리 (25. 11.04 작업 내역)

   1. 메시지 내역 조회 API (`ChatMessageListView`) 버그 수정:
       * ChatRoomService.get_chat_rooms_for_user 메서드에서 발생하던 ProgrammingError: more than one row returned by a subquery used as an expression 오류를 해결했습니다. unread_count 계산 로직을 OuterRef와 Coalesce(Subquery(...), Value(0))를 사용하여 정확하고 효율적으로 재작성
       * ChatMessageListView 내의 user.is_authenticated 중복 검사 로직을 제거하여 코드 가독성을 높이고 불필요한 코드 경로를 제거

   2. 테스트 커버리지 개선:
       * apps/chat/views.py 파일의 테스트 커버리지를 100%로 달성
       * 페이지네이션이 비활성화된 경우의 응답을 테스트하는 test_message_list_no_pagination 테스트 케이스를 test_chat_api.py에 추가
       * apps/chat/tests/test_chat_room_api.py의 test_get_chat_room_list_unauthenticated 테스트에서 이전 인증 상태가 유지되는 문제를 해결하기 위해 self.client.logout()을 추가

   3. `daphne` 의존성 제거 및 `ChatConsumer` 테스트 리팩토링:
       * test_chat_consumers.py 파일이 daphne 라이브러리에 의존하여 테스트가 실패하던 문제를 해결
       * AsgiTestCase 대신 APITestCase를 상속받고 WebsocketCommunicator를 직접 사용하여 ChatConsumer를 테스트하도록 코드를 리팩토링
       * 비동기 테스트 메서드 내에서 동기 ORM 호출(create_user, create)을 asgiref.sync.sync_to_async로 감싸 AttributeError를 해결
       * WebsocketCommunicator 사용 시 scope에 url_route 정보가 자동으로 추가되지 않아 발생하던 KeyError를 해결하기 위해, 테스트 코드에서 communicator.scope["url_route"]를 수동으로 설정
       * 이 리팩토링을 통해 daphne 개발 의존성을 제거하고, channels 공식 권장 테스트 방식 준수

  추가 정보:

   * consumers.py 파일의 커버리지는 현재 93%
   * 내일 웹소켓 라우팅 작업 시 추가 테스트를 통해 향상 예정

   * apps/chat/__init__.py: 100%
   * apps/chat/apps.py: 100%
   * apps/chat/consumers.py: 93% (누락 라인: 28-29, 65)
   * apps/chat/models/__init__.py: 100%
   * apps/chat/models/chat_message.py: 100%
   * apps/chat/models/last_read_message.py: 100%
   * apps/chat/pagination.py: 100%
   * apps/chat/serializers.py: 100%
   * apps/chat/services/__init__.py: 100%
   * apps/chat/services/chat_room_service.py: 100%
   * apps/chat/views.py: 100%

## 📸 스크린샷 (선택)

## 📝 기타 참고 사항
•	기존 36ca04d 커밋에서 진행된 '메시지 내역 조회 기능 구현 및 테스트 커버리지 향상' 작업
•	프론트엔드 요청사항 반영 : 기본키로 아이디 지정을 했으면 프론트엔드에서는 사용자 구분이 어려움 -> 메시지 내역 조회에도 sender_id 뿐만 아니라 sender_nickname도 추가가 필요 -> 반영 완료
•	코치님 피드백 반영: 파일전송 관련 필드 제거, message가 fk로 결합되어 있지않아 message_id로 변경, 350개의 쿼리 발생 -> 1개쿼리, content / sender_nickname / created_at 으로 정리
•	채팅 기능 테스트 커버리지 개선 작업 중 예상치 못한 상황이 발생

【발견한 문제】
- ChatConsumer의 receive 메서드에 버그 발견
- 기존 테스트가 daphne 의존성 문제로 불완전하게 동작

【해결 과정】
1. 버그 수정 및 테스트 추가 시도
2. Django Channels 공식 권장 방식(WebsocketCommunicator)으로 테스트 리팩토링
3. daphne 의존성 완전 제거
4. daphne 의존성 문제로 project.toml 파일에 의존성 추가 -> poetry.lock 업데이트

【결과】
✅ chat 앱 커버리지: 73% → 98%
✅ 테스트 안정성 개선 (업계 표준 방식 적용)
✅ 불필요한 의존성 제거

【PR 분리 불가 이유】
- 작업들이 모두 "테스트 품질 개선"으로 연결되어 있음
- 중간 단계로 PR 분리 시 CI/CD 커버리지 검사 통과 불가

변경 라인이 350라인을 넘게 되었는데.. 하나의 맥락으로 묶인 작업이라 단일 PR로 제출하고자 합니다.
대형 PR 규칙에 위반되었지만 분할하면 커버리지가 통과가 안되고 있는 상태였습니다.

## 🧪 PR Checklist
- [x] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [x] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

BE12-OZ committed:
✨ feat: 메시지 목록 조회 API 구현 및 읽음 처리 로직 추가

- `apps/chat/serializers.py`: `ChatMessageSerializer`에 `is_read` 로직 구현 (LastReadMessage 활용).
- `apps/chat/pagination.py`: `ChatMessagePagination` 클래스 정의 (첫 페이지 300개, 이후 100개 메시지).
- `apps/chat/views.py`: `ChatMessageListView`에 `ChatMessagePagination` 적용 및 멤버 권한 확인 로직 추가.
- `apps/chat/urls.py`: 메시지 목록 조회 API 엔드포인트 추가.
- `config/urls.py`: `chat` 앱 URL 포함.
- `apps/chat/consumers.py`: `connect` 메서드에 메시지 읽음 처리 로직 추가 (LastReadMessage 업데이트).
- `apps/chat/tests/test_chat_api.py`: 메시지 목록 조회 API에 대한 통합 테스트 코드 작성.

# Conflicts:
#	apps/chat/consumers.py

# Conflicts:
#	apps/chat/pagination.py
#	apps/chat/serializers.py
#	apps/chat/urls.py
#	apps/chat/views.py

BE12-OZ committed:
refactor: feature/chat-message-list-new 브랜치에서 중복된 읽음 처리 로직 제거

develop 브랜치에 이미 올바른 읽음 처리 로직이 병합되었으므로, feature/chat-message-list-new 브랜치의 consumers.py 파일에서 중복되고 버그가 있는 읽음 처리 로직과 관련 import 문을 제거합니다.

BE12-OZ committed:
✨ feat: 메시지 내역 조회 API 구현 및 읽음 처리 로직 추가

- 메시지 내역 조회 API (/api/v1/study-groups/{study_group_id}/messages) 구현.
- 페이지네이션 (최초 300개, 무한 스크롤 100개) 적용.
- API Spec (extend_schema) 상세화.
- 인증 및 그룹 멤버십 기반 보안 강화.
- is_read 상태 반환 로직 포함.
- 모든 테스트 통과 및 높은 코드 커버리지(92%).
- mypy, black, isort 검사 통과.

BE12-OZ committed:
✨ feat: 메시지 내역 조회 API 구현 및 읽음 처리 로직 추가

•메시지 내역 조회 API (GET /api/v1/study-groups/{study_group_id}/messages) 구현

•페이지네이션(초기 300개, 이후 무한스크롤 시 100개) 적용

•extend_schema로 API Spec 상세화 및 응답 스키마 명확화

•인증 및 스터디 그룹 멤버십 기반 권한 검증 추가

•메시지별 is_read 상태 반환 로직 포함

•모든 테스트 통과, 코드 커버리지 92%

•mypy 타입 오류 수정 및 테스트 통과

•`pyproject.toml`에 `mypy` 설정을 추가하여 `channels` 라이브러리의 `import-untyped` 오류를 해결

BE12-OZ committed:
✨ feat: chat 앱 모델 및 시리얼라이저 테스트 커버리지 향상

### 변경 사항 요약

 앱의 모델( 메서드) 및 시리얼라이저(, , )에 대한 테스트 커버리지를 90% 이상으로 향상시켰습니다.

### 상세 변경 내역

-   ** (시리얼라이저):**
    -   와 에 대한 테스트 케이스를 전면적으로 추가하여 직렬화/역직렬화 로직의 정확성을 검증했습니다.
    -   의  및  메서드에 대한 엣지 케이스 테스트를 추가하여 커버리지를 **91%**로 향상시켰습니다.
-   ** (모델):**
    -    모델의  메서드에 대한 테스트 케이스를 추가하여, 가 있는 경우와 없는 경우 모두 올바른 문자열 표현을 반환하는지 확인하여 커버리지를 **100%**로 향상시켰습니다.
-   ** (모델):**
    -    모델의  메서드에 대한 테스트 케이스를 추가하여 올바른 문자열 표현을 반환하는지 확인하여 커버리지를 **100%**로 향상시켰습니다.

BE12-OZ committed:
feat(chat): 채팅 앱 메시지 내역 API 구현, 버그 수정 및 테스트 커버리지 개선

BE12-OZ committed:
fix: Chat 앱 mypy 오류 및 테스트 어설션 수정

채팅 애플리케이션의 두 가지 문제, mypy 유형 검사 오류와 잘못된 테스트 어설션을 해결

1.  **`apps/chat/services/chat_room_service.py`의 Mypy 오류**:
    *   **문제**: 사용자 멤버십으로 `StudyGroup`을 필터링할 때 Mypy가 "Cannot resolve keyword 'group_members'" 오류를 표시
    *   **해결**: `get_chat_rooms_for_user`를 리팩토링하여 `GroupMember`에 대한 서브쿼리를 사용하여 `study_group` ID를 식별

2.  **`apps/chat/tests/test_chat_room_api.py`의 테스트 실패**:
    *   **문제**: `test_get_chat_room_list_success`가 실패하여 `user1`에 대해 채팅방을 2개만 발견 -> `setUp` 메서드는 `user1`을 두 개의 `StudyGroup` 인스턴스에만 올바르게 추가
    *   **해결**: 테스트 설정과 일치하도록 어설션을 `self.assertEqual(len(response.data), 3)`에서 `self.assertEqual(len(response.data), 2)`로 수정

**추가 변경 사항**:
*   `apps/chat/tests/test_chat_api.py`: 테스트 중 순진한 datetime으로 인한 `RuntimeWarning`을 해결하기 위해 `setUp`에서 `StudyGroup` 생성 시 `timezone.make_aware(datetime(...))`을 사용하도록 업데이트
*   `poetry.lock` 및 `pyproject.toml`: `poetry run` 명령으로 인한 종속성 변경으로 업데이트

**검증**:
`chat` 앱에 대한 모든 로컬 검사가 통과했습니다: 코드 포맷팅, mypy 유형 검사, 단위/통합 테스트.

2joonkim commented on apps/chat/services/chat_room_service.py:
```diff
@@ -2,53 +2,50 @@
-            last_message_created_at=Subquery(last_message_subquery.values("created_at")),
-            unread_count=Coalesce(Subquery(unread_count_subquery, output_field=models.IntegerField()), 0),
+        # 최종적으로 필요한 값들을 annotate
+        chat_rooms = groups_with_last_read.annotate(
```
이 부분 공통 서브쿼리로 하나 정의해두고
```
last_message_subquery = ChatMessage.objects.filter(
    study_group=OuterRef("pk")
).order_by("-created_at")
```

각 필드별로 재활용해서 쓰는게 더 좋아보이는데 지금 기준으로만 봤을 때는 각 필드마다 서브쿼리문을 쓰시니까
안좋을 거 같은데 한번 변경해보시죠

Meoyoug commented on apps/chat/services/chat_room_service.py:
```diff
@@ -2,53 +2,50 @@
         # 사용자가 속한 스터디 그룹 목록 조회
-        study_groups = StudyGroup.objects.filter(members__user=user)
+        member_study_group_ids = GroupMember.objects.filter(user=user).values("study_group")
+        study_groups = StudyGroup.objects.filter(id__in=Subquery(member_study_group_ids))
```
이렇게 바꾼 이유가있을까요?

Meoyoug commented on apps/chat/serializers.py:
```diff
@@ -1,27 +1,78 @@
+            last_read_message = LastReadMessage.objects.get(user=request.user, study_group=obj.study_group)
+            return obj.id <= last_read_message.message_id
+        except LastReadMessage.DoesNotExist:
+            return False  # No last read message for this user in this group
```
이거 annotate + subquery 조합으로 해결할 수 있었을 것 같은데용?

Meoyoug commented on apps/chat/views.py:
```diff
@@ -97,12 +97,10 @@ def get_queryset(self) -> QuerySet[ChatMessage]:
     def get(self, request: Request, *args: Any, **kwargs: Any) -> Response:
         study_group_id = self.kwargs["study_group_id"]
         user = request.user
+        assert user.is_authenticated
```
authentication_classes 와 permission_classes에 의해서 유저의 로그인 상태가 검증된다면 assert대신 cast를 활용해보시면 좋을 것 같습니다.

BE12-OZ committed:
feat(chat): Implement chat message listing and related services\n\n- Refactor chat services into chat_service.py\n- Add permissions for chat access\n- Update serializers, views, and URLs for message listing\n- Remove old chat service files and serializers

BE12-OZ committed:
fix(chat): 채팅 테스트 및 mypy 오류 수정

원인:
- `apps/chat/pagination.py`에서 `Paginator` 타입 힌트가 런타임 `TypeError`를 유발하고 `mypy` 오류를 발생시켰습니다.
- `apps/chat/tests/test_chat_consumers.py`에서 `UserManager.acreate_user` 호출로 인한 `mypy` 오류가 발생했습니다.
- `apps/chat/tests/test_chat_consumers.py`에서 `KeyError: 'user'` 및 `KeyError: 'url_route'`가 발생했습니다.
- `apps/chat/tests/test_chat_consumers.py`에서 `AssertionError: 403 != '403'` 타입 불일치 오류가 발생했습니다.
- 비동기 테스트 환경에서 데이터베이스 연결이 조기에 닫히는 문제가 발생했습니다.

해결:
- `apps/chat/pagination.py`의 `Paginator` 타입 힌트를 런타임 오류를 피하고 `mypy`를 통과시키기 위해 `paginator: Paginator`로 변경하고 `# type: ignore[type-arg]` 주석을 추가했습니다.
- `apps/chat/tests/test_chat_consumers.py`에서 `User.objects.acreate_user` 호출을 `self._create_user` 헬퍼 메서드 사용으로 변경했습니다.
- `apps/chat/tests/test_chat_consumers.py`의 `test_connection_failure_invalid_study_group` 테스트에 누락된 `communicator.scope["user"]` 및 `communicator.scope["url_route"]` 설정을 추가했습니다.
- `apps/chat/tests/test_chat_consumers.py`의 연결 실패 테스트들에서 `subprotocol` 어설션을 문자열 `'403'`에서 정수 `403`으로 변경하여 타입 불일치 오류를 수정했습니다.
- `apps/chat/tests/test_chat_consumers.py`의 `ChatConsumerTest` 기본 클래스를 `APITestCase`에서 `TransactionTestCase`로 변경하여 비동기 테스트 환경에서 데이터베이스 연결 문제를 해결했습니다.
